src/charm.py

```diff
@@ -174,17 +138,23 @@ def __init__(self, *args):
         self.ingress_per_appv2 = ipa_v2 = IPAv2(charm=self)
 
         self.ingress_per_unit = IngressPerUnitProvider(charm=self)
+
         self.traefik_route = TraefikRouteProvider(
-            charm=self, external_host=self.external_host, scheme=self._scheme  # type: ignore
+            charm=self, external_host=self._external_host, scheme=self._scheme  # type: ignore
+        )
+
+        self.traefik = Traefik(
+            _TRAEFIK_SERVICE_NAME,
+            self.container,
+            routing_mode=self._routing_mode,
+            tcp_entrypoints=self._tcp_entrypoints(),
+            tls_enabled=self._is_tls_enabled(),
         )
 
-        web = ServicePort(self._port, name=f"{self.app.name}")
-        websecure = ServicePort(self._tls_port, name=f"{self.app.name}-tls")
-        tcp_ports = [ServicePort(int(port), name=name) for name, port in self._tcp_ports.items()]
         self.service_patch = KubernetesServicePatch(
             charm=self,
             service_type="LoadBalancer",
-            ports=[web, websecure] + tcp_ports,
+            ports=self.traefik.service_ports,
```

Comment:
Per other comment, I think the k8s datastructure should be constructed here in charm.py, not there.
Could still use e.g. `self.traefik.port_x`.

Reply:
mm, that does feel like a charm concern, indeed.